### DISCUSSION TITLE
Prevent saving while falling

### DISCUSSION THREAD

**Discussion by TheStolenBattenberg** [stohrendorf/CroftEngine#288] [General]
It's habit after a while to spam through the menus, and it's very easy to accidently hit right and end up in the save menu (which looks identical to the load menu) As it currently stands, you can save while falling to your death and totally ruin your save file (if you're not savvy with editing them, or using multiple slots) - it'd be nice to stop this so you don't have to manually fix your save file in the event of accidently saving while Lara falls to her death...

**Comment by stohrendorf** [comment]
IMHO, it's an essential part of Tomb Raider to feel the accidents you made yourself. Although I think it would be an interesting feature, I think it would ruin the experience, as TR1 is basically somewhat of a survival game. I can understand it's not "easy", but being "easy" is - to my understanding - not the spirit of the game. I.e., doing "accidents" is part of survival. This is why I decided against making it _too_ easy. It's an totally different issue concerning accesibility, but as far as I'm aware, that wasn't an issue before.

**Comment by TheStolenBattenberg** [comment]
While I do agree with your argument that punishment is certainly part of Tomb Raider - watching the same fall _X_ times in a row is tiring, and if you've got a limited time for playing games (got children myself, so at most one hour a day while they nap) , the deaths all add up and limits you further... Surely if you wanted to some sort of hardcore, original experience - it would be better to re-implement the save crystals of the PSX release, and allow that as an optional layer of difficultly like games such as Legend of Grimrock did with their optional Iron Man or 'single use crystals' modes.

To me personally (and I don't say this without experience as I have my own projects of a similar type: open source ports, game development and QA) - this just feels like a massive oversite by Core Design. It's softlocking your save, potentially wasting hours of our actual finite life that have been put into the save at that point, and that's not really an acceptable punishment since it isn't immersive in the slightest - just disheartening.

I do agree that this was never an issue before - I never ran into it while playing and completing any of the TR games previously, but I'm trying out Sabatu's TR1 - and that has some seriously nasty requirements for good acrobatics. If I were to play without saving before a complicated jumping section, I'd wager on it taking years for me to complete.

TLDR; To me at least this feels like a bug, and it wouldn't be too hard to prevent from happening.

P.S: I appreciate your work with CroftEngine - I've been waiting a long time to revist TR1, but the lead developer of OpenTomb seemingly refused to continue any work, and seemingly all other projects end up being either on the back burner or discontinued too.

**Comment by stohrendorf** [comment]
Hm. Ok. Then, please tell me a _certain_ condition to allow (or disallow) saves. I had a few thoughts about that, but nothing was "enough". If you have some really convincing idea, please tell me.

**Comment by TheStolenBattenberg** [comment]
I could see this working by checking if Lara is on the ground to enable saving, or, if she's falling to disable saving. I'm aware there's other ways to potentially soft lock yourself like this, but they're a lot harder to achieve, and it's possible to survive (most of) them. I'm going to list them regardless in order to work out which conditions would be needed and when you'd want to prevent saving.

- _A close enemy with low health_: You can always likely just open the menu to use a medipack.
- _Lara is on fire_: You can spam medipacks until you reach a body of water. I'm pretty sure there's always a body of water somewhere in a level where there is fire.
- _Lara is drowning_: Again, medipack spam makes it possible to avoid this.
- _Falling_: This is pretty likely to break your save.
- **_Saving while you are dead_**: This WILL break your save, but is a little diverted from the point of this issue - it is pretty funny that it's possible, however.

I can't recall anymore likely occurrences, but in conclusion I believe it would be best to disallow saving on the conditions of the Lara not being _on solid ground_ or _being dead_. Anything else is potentially avoidable, and a lot less likely than these two.

Another potential solution to this, which would blanket cover all the potential soft locking, and which allows both our opinions of difficultly would be to just automatically move the save slot cursor to the next free slot when you go to save the game. That would solve the issue just as well as an addition of checks to prevent saving. It does add the requirement of more storage space for save data, but if someone were to care about that I still think it's preferable to move the cursor up by one than to accidently ruin your save game... Aside from that, modern hardware does tend to have a lot of storage space at its disposal.

**Comment by stohrendorf** [comment]
I tend to add a limited "quicksave history" rather than adding "save guards". Given that there is a total of 100 save slots, it would boil down to management, and the biggest issue is - to my understanding - that there's currenly only a single quicksave slot.

**Comment by TheStolenBattenberg** [comment]
Maybe I should clear up a potential miss-understanding? I was talking about the actual save slots the whole time, rather than quicksaves. I always played TR on PS1 or emulated... Thus, I'm accustomed to a routine of "open menu, hit down, press x, hit right and double tap x" for save with muscle memory - don't hit right for load which is where the accident happens. TR2 at least, doesn't let you save while you're dead because it automatically opens the passport to the Load Game screen. Not sure about TR1.

If there is an actual quicksave history implemented at the moment, maybe this isn't actually a problem?.. And my apologies if that is the case.

**Comment by stohrendorf** [comment]
You can't save when dead, so I assumed that statement was written by mistake. Also, as of right now, there's no quicksave history implemented. I'm sorry to ask, but have you actually used the engine, or have you brought this up just because you have this issue with every other engine and assume it's also the case for this one?

**Reply by TheStolenBattenberg** [reply]
As I said previously, I'm using it right now to play Sabatu's Tomb Raider I reimagining ([proof if you really need to verify I am indeed using CroftEngine](https://pasteboard.co/EzYUg4uoxuWm.png)).

You actually can save while you're dead - I haven't gone through the code for for CroftEngine to see exactly how death works here, but I'd assume that you either haven't actually set a 'dead state' until the passport appears, at which point indeed you cannot save... However while waiting for the passport, you can hit start (on a gamepad) and save. [Here's a 7Z file of a saved death (GitHub wouldn't let me attach it here).](https://mega.nz/file/3b5U0Z4a#SFul2YQ5ymkwfzcoyvST8hmZ25uIcH_OrlY3PQQ3F3E)

**Reply by stohrendorf** [reply]
My bad, sorry, for some reason that statement didn't stick in my mind. I didn't want to accuse you or anything like that.

Anyways, I'm a bit surprised you can open the passport while dead and actually save there. Seems to be an oversight when guarding the save logic.

Regarding all the conditions you listed when it is safe to save and when it is not...
* "A close enemy with low health" - your solution assumes that a medi pack is available, and if you don't have any, what to do then? Do you disallow saves at all, or do you disable the safety checks alltogether? Also, what is considered "low"? Do you take your weapon damage compared to the baddie's health into account?
* "Lara is on fire" - even while you _could_ use up hundreds of medi packs if you had them, you're seriously hurt when you don't have any. Disabling saves here seems to be the best solution.
* "Lara is drowning" - that's most likely the biggest issue. Do you allow saving when you're close to the surface while drowning? How close is to be considered a "safe distance" to allow saving? Every time you start diving, there's the potential that you don't find back to the surface, so it's never safe to save when underwater, and that would effectively mean you had to disable saves when underwater. I don't think that would be such a good idea when you lose up to 2 minutes of gameplay in a larger area just because you messed something small up just before reaching the surface.
* "Falling" - what about when you will do a safe landing? And what if you save while standing on a timed trapdoor? That last one is especially tricky, because you can't reliably determine without massive simulations that the door will stay closed. For example, consider a timed trigger that triggers a boulder, which in turn rolls over a timed floor trigger, which then in turn triggers the trapdoor.

I want to suggest another solution: re-train you muscle memory to stop when you're asked to overwrite a slot ;)

**Reply by stohrendorf** [reply]
Or, another solution could be to optionally add a 3 second delay before allowing to confirm the overwrite.

**Reply by TheStolenBattenberg** [reply]
> My bad, sorry, for some reason that statement didn't stick in my mind. I didn't want to accuse you or anything like that.

No problem, I understand why you asked. :)

Taking in account what you've said about the conditions I listed, you're probably right about me just retraining my muscle memory - it is pretty hard to find a good solution for any of the possible conditions as you pointed out and making assumptions about the condition of the players inventory isn't a great idea. Having said that, I still believe that increasing the save slot number to the next free slot would be a great solution (even if it was optional, via a configuration option or something), or as you suggested a quicksave history would be good too (for anyone using that.) I myself am not keen on a delay, waiting around for prompts is generally very boring.

I've noticed yesterday that you declared in the repository read me that you're only offering support for CroftEngine now, so if that's truly the case I don't believe a solution for 'saving while you're about to die' needs to be implemented; though, it would be nice to see a fix for saving while Lara is dead, even if it is sort of morbidly funny.

**Reply by stohrendorf** [reply]
There are a few possible solutions:
* add an option to either select the most recent save slot, or to select the first free one after it
* adding a delay

About the save issue while being dead, that's certainly a bug and needs to be fixed.

**Reply by stohrendorf** [reply]
@TheStolenBattenberg I've got a question: considering the scenario the engine would automatically choose the next free save slot after the most recent save, what should happen if there is no free slot available?

**Reply by TheStolenBattenberg** [reply]
From what I can see in croft engine, there's 10 pages of 10 slots? I'd say that there has to be limits to what can be achievable as a safe guard.  At some point you gotta call it a day and say 'this is good enough to prevent at least some of the accidents." If the end user gets to the point where they'ved used all 100 possible slots, they should be cleaning up some of the slots they're not using, until they do the game should go back to overwriting the last slot it was able to save into.

The only other solution would be to implement theoretically infinite slots.

**Reply by stohrendorf** [reply]
The internal issue numbers for this are CE-599, CE-596, CE-597

**Reply by stohrendorf** [reply]
The following features will be in the next release:
* option to require you to hold down the button when confirming overwrites for a few seconds (i.e., you must press "yes" for a few seconds before it gets accepted if you enable this)
* option to select the first free or oldest slot on save instead of the most recent one
* a lot of save cleanup actions by using the secondary menu action in the saves list
